docs/rfc/rfc-017-abci++-vote-extension-propag.md

```diff
@@ -0,0 +1,438 @@
+up to date.
+Thus, blocksync has all the data it requires to switch to the consensus reactor, as long as:
+
+- The node is still at height 0 (where no commit or extended commit is needed)
+- The node has processed at least 1 block in blocksync
+
+Additionally, when a validator falls behind while having already switched to the consensus reactor,
+peer nodes have enough information in their block store to reconstruct the extended votes and send
+them to the validator falling behind.
+
+This solution requires a few changes to the consensus reactor:
+
+- upon saving block for a given height in the block store at decision time, save the extended commit
+  as well
+- in the catch-up mechanism, when a node realizes that another peer is more than 2 heights
+  behind, it uses the extended commit (rather than the canoncial commit as done previously) to
+  reconstruct the votes with their corresponding extensions
+
+The changes to the blocksync reactor are more substantial:
+
+- the `BlockResponse` message is extended to include the extended commit
+- structure `bpRequester` is likewise extended to hold the received extended commits coming in
+  `BlockResponse` messages
+- method `PeekTwoBlocks` is modified to also return the extended commit corresponding to the first block
+- when successfully verifying a received block, the reactor saves its corresponding extended commit in
+  the block store
+
+The two main drawbacks of this base implementation are:
+
+- the increased size taken by the block store, in particular with big extensions
+- the increased bandwith taken by the new format of `BlockResponse`
+
+#### Possible Optimization: Pruning the Extended Commit History
+
+If we cannot switch from the consensus reactor back to the blocksync reactor, and we do not have
+the catch-up mechanism in the consensus reactor explained above, we cannot prune the extended commit
+backlog in the block store without sacrificing the implementation's correctness. The asynchronous
+nature of our distributed system model allows a process to fall behing an arbitrary number of
+heights, and thus all extended commits need to be kept *just in case* a node that late had
+previously switched to the consensus reactor.
+
+However, there is a possibility to optimize the base implementation. We could prune from the block
+store all extended commits that are more than *d* heights in the past. Then, we need to handle two
+new situations, roughly equivalent to cases (g.1) and (g.2) described above.
+
+- (g.1) A node starts from scratch or recovers after a crash, and (for simplicity of explanation)
+  it does not run statesync.
+  In this case, the blocksync reactor needs to be improved to make sure it cannot switch to consensus
+  until it has received a valid extended commit for a height in its future. This extended commit
+  will allow the node to switch from the blocksync reactor to the consensus reactor and immediately
+  act as a proposer if required.
```

Comment:
I think that this can somewhat be extended to mean to any type of 'catch-up' mode.

We realistically have two modes:
1. Catchup
2. Active Consensus

In catchup, as you describe above, you _only_ receive `FinalizeBlock` requests. You don't have the latest `VoteExtensions` so you're not actively participating in consensus and you do not have enough data to switch to participating. 

In Active Consensus, you have the vote extension data and are therefore allowed to propose and participate. It seems that, whatever mechanism we take to fully implement the feature, it looks like we need to gate switching from Catchup -> ActiveConsensus on whether or not you have the vote extension data. One way to implement 'catchup' is by allowing the node to bypass proposal and another way is by moving in-and-out of blocksync.

Reply:
Hmmm, the cases 1. and 2. you are describing are the (g.1) and (g.2) in the text.
If you didn't recognize that, I might need to improve the text. I guess it's best to discuss in person how to go about this.

Reply:
Got it, to a degree yes. I'm somewhat elaborating that no matter the mechanism, blocksync and consensus where proposing is skipped, are both effectively cases of 'catchup'. So, the overall challenge in front of us is both ensuring that we never go from 1 -> 2 without enough vote extensions and we never become stuck unable to go from 1 -> 2.

Reply:
Yes, exactly. That is, in essence what the property at the core of Solution 3 ensures (described in abstract terms).
For our implementation, we can't do it _as such_ until we can switch back and forth. Until then, we are forced to keep a certain backlog of extended commits (rather than just 1 or 2 as in Solution 3)